### PR TITLE
SCP-5019 Coin selection respects collateral rules

### PR DESCRIPTION
**Pull request by bwbush**
This PR makes `selectCoins` conform the the `WalletContext` requirement that it uses the only the specified collateral UTxOs if any are specified.

1. Collateral is only selected if a Plutus script is being run.
2. Collateral may be any of the wallet's UTxOs if collateral is not explicitly specified.
3. Collateral may only be specified collateral UTxOs if any are specified.
4. Coin-selection collateral unit test includes new features:
    - Collateral selection if and only if a Plutus script is run.
    - Collateral must be one of the eligible UTxOs for collateral (see 2 and 3 above)
    - Wallet may or may not have specified explicit collateral. 

Tested with 1M quickcheck test cases.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [ ] [Test report is updated](https://github.com/input-output-hk/marlowe-cardano/blob/main/marlowe/test/test-report.md) (if relevant)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs [hunk 1]
```diff
@@ -207,6 +213,47 @@ spec = do
...
+          , (20, hedgehog $ do
+                txIn <- genTxIn
+                script <- PReferenceScript <$> genTxIn <*> pure Nothing
+                datum <- ScriptDatumForTxIn <$> genScriptData
+                redeemer <- genScriptData
+                pure (True, emptyTxBodyContent {txIns = [(txIn,
+                  BuildTxWith
+                    . ScriptWitness ScriptWitnessForSpending
+                    $ PlutusScriptWitness PlutusScriptV2InBabbage PlutusScriptV2
+                        script datum redeemer (ExecutionUnits 0 0))]}))
```
Comment (palas) [inline marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs:254]: What do these cases represent? What is the diference between `PReferenceScript` and `SReferenceScript`

**Review by palas — APPROVED** [review]
It looks good to me. But I am missing quite a lot of context

File: marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs (hunk 1 above)
Comment (bwbush) [reply marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs]: `PReferenceScript` and `SReferenceScript` are Plutus and Simple reference scripts, respectively.

I added comments to show what the cases represent, but they are essentially situations that need or don't need collateral: https://github.com/input-output-hk/marlowe-cardano/blob/6c174a39ef200bf67e0901c7e4ea6de28f06a325/marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs#L219-L259

Comment (palas) [reply marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs]: Thanks

**Pull request merged**